File: engine/core/base/engine.cpp
// engine/core/base/engine.cpp
#include "engine.hpp"

namespace r2d2 {

    Engine::Engine() : running(false) {}

    Engine::~Engine() {
        if (running) shutdown();
    }

    void Engine::init() {
        running = true;
    }

    void Engine::shutdown() {
        running = false;
    }

    void Engine::tick() {
        if (!running) return;
        // future: events, lifecycle, renderer, input...
    }

    bool Engine::is_running() const {
        return running;
    }

}

File: engine/python/engine/engine.py
# engine/python/engine/engine.py

from engine.cython.base.engine import PyEngine

class Engine:
    def __init__(self):
        self._engine = PyEngine()

    def init(self):
        self._engine.init()

    def shutdown(self):
        self._engine.shutdown()

    def tick(self):
        self._engine.tick()

    def is_running(self):
        return self._engine.is_running()

    def run(self):
        self.init()
        while self.is_running():
            self.tick()
